### PR TITLE
Network Router: More generic use of SupportsFeatureMixin

### PR DESCRIPTION
**Pull request by gildub**
Migrated from https://github.com/ManageIQ/manageiq/pull/13073

### CODE REVIEW

File: app/models/manageiq/providers/openstack/network_manager/network_router.rb [hunk 1]
```diff
@@ -2,10 +2,26 @@ class ManageIQ::Providers::Openstack::NetworkManager::NetworkRouter < ::NetworkR
...
-  supports :create_network_router
-  supports :delete_network_router
-  supports :update_network_router
   supports :add_interface
+
+  supports :create
+
+  supports :delete do
+    if ext_management_system.nil?
+      unsupported_reason_add(:delete_network_router, _("The Network Router is not connected to an active %{table}") % {
```
Comment (durandom) [inline app/models/manageiq/providers/openstack/network_manager/network_router.rb]: the reason should match the one in the `supports` block.
` unsupported_reason_add(:delete ...`

File: app/models/manageiq/providers/openstack/network_manager/network_router.rb [hunk 2]
```diff
@@ -2,10 +2,26 @@ class ManageIQ::Providers::Openstack::NetworkManager::NetworkRouter < ::NetworkR
...
+        :table => ui_lookup(:table => "ext_management_systems")
+      })
+    end
+  end
+
+  supports :update do
+    if ext_management_system.nil?
+      unsupported_reason_add(:update_network_router, _("The Network Router is not connected to an active %{table}") % {
```
Comment (durandom) [inline app/models/manageiq/providers/openstack/network_manager/network_router.rb]: same here

File: app/models/manageiq/providers/openstack/network_manager/network_router.rb (hunk 1 above)
Comment (gildub) [inline app/models/manageiq/providers/openstack/network_manager/network_router.rb]: Absolutely

File: app/models/manageiq/providers/openstack/network_manager/network_router.rb (hunk 2 above)
Comment (gildub) [inline app/models/manageiq/providers/openstack/network_manager/network_router.rb]: Same

**Comment by gildub** [comment]
@durandom, to be consistent with other network item I'm just using the generic action.

**Comment by tzumainn** [comment]
Is there an associated PR to manageiq-ui-classic?  I think the buttons there check for things like :create_network_router.

**Comment by aufi** [comment]
Tested, I wasn't able see error (router disappeared together with network manager deletion), but anyway merging should not cause errors, trust @durandom ;-)

**Comment by gildub** [comment]
@tzumainn, 

You're right we need to clean up the associated declaration, I've created 
https://github.com/ManageIQ/manageiq/pull/15451 to take care of it.

Thanks

**Comment by gildub** [comment]
@aufi, thanks for the confirmation.

btw, would you mind detailing here or by email more about the "router disappeared together with network manager" ?

**Comment by aufi** [comment]
@gildub Sure, I understood this change as a check if Router is associated with provider (network manager) and if not, disable update, delete action because the action couldn't be completed without valid provider.

The only case when we have an object in MIQ database which doesn't belong to any existing provider I'm aware of is case when we _delete the provider_, but some of its inventory remain in MIQ database (e.g. Host in Infra provider). I expected that Router has similar behaviour, so it stays in MIQ database event its provider (network manager) was deleted and change in this PR ensures, that such router cannot be edited or deleted.

I might be wrong and checking of valid provider is just a convention in ManageIQ providers or feel free correct me, thanks :-)

**Comment by gildub** [comment]
@aufi, 

I see, well that's an interesting behaviour, I'd assume the dependent objects of a provider to clear up (cascading delete) no matter what, if that's not the case then we are in the *wild* ;).

**Comment by aufi** [comment]
@gildub Yeah, in most cases there is the cascade delete, no problem.

**Comment by mansam** [comment]
@miq-bot add_label fine/yes

**Comment by gildub** [comment]
Fine branch is taking place within manageiq repo. The cherry pick for this is patch has been created as https://github.com/ManageIQ/manageiq/pull/17660

**Comment by gildub** [comment]
Never mind my previous comment as the cherry pick can be done across.

**Comment by simaishi** [comment]
Fine backport (to manageiq repo) details:

```
$ git log -1
commit 9cfc0ea4c6da05bd928576d6984e1662ed9e568f
Author: Marek Aufart <aufi.cz@gmail.com>
Date:   Mon Jun 26 11:58:30 2017 +0200

    Merge pull request #56 from gildub/network_controller_network_router_fix_support_feature
    
    Network Router: More generic use of SupportsFeatureMixin
    (cherry picked from commit 71520a1bced4e606b6703631cf40093044ef1838)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1593349
    https://bugzilla.redhat.com/show_bug.cgi?id=1593353
```

**Comment by simaishi** [comment]
@gildub Can you please take a look at manageiq-ui-classic Travis failure for Fine branch?
https://travis-ci.org/ManageIQ/manageiq-ui-classic/jobs/421223743

I suspect this PR is causing the failure...

```
  1) ApplicationHelper::Button::NetworkRouterNew#disabled? when at least one provider supports network router create then the button is not disabled
     Failure/Error: expect(button.disabled?).to be false
     
       expected false
            got true
     # ./spec/helpers/application_helper/buttons/network_router_new_spec.rb:7:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:75:in `block (3 levels) in <top (required)>'
     # ./spec/manageiq/spec/support/evm_spec_helper.rb:34:in `clear_caches'
     # ./spec/spec_helper.rb:75:in `block (2 levels) in <top (required)>'
  2) ApplicationHelper::Button::NetworkRouterNew#calculate_properties when at least one provider supports network router create, the button has no error in the title
     Failure/Error: expect(button[:title]).to be nil
     
       expected nil
            got #<String:474129520> => "No cloud providers support creating network routers."
     # ./spec/helpers/application_helper/buttons/network_router_new_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:75:in `block (3 levels) in <top (required)>'
     # ./spec/manageiq/spec/support/evm_spec_helper.rb:34:in `clear_caches'
     # ./spec/spec_helper.rb:75:in `block (2 levels) in <top (required)>'
```

**Comment by simaishi** [comment]
@gildub ping ^

**Comment by gildub** [comment]
@satoe,

I'll look at it as soon as possible. BTW don't we need to restore fine/yes flag?

**Comment by simaishi** [comment]
@gildub Thanks. Since this PR is backported, it can stay as `fine/backported`. Please reference this PR in your PR that will fix the Travis failure though.

**Pull request merged**